<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Apply Poisson thinning to a matrix of count data. — poisthin • seqgendiff</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Apply Poisson thinning to a matrix of count data. — poisthin"><meta property="og:description" content="This is now defunct. Please try out select_counts and
thin_2group."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/seqgendiff/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Apply Poisson thinning to a matrix of count data.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/seqgendiff/blob/HEAD/R/poisthin.R" class="external-link"><code>R/poisthin.R</code></a></small>
    <div class="hidden name"><code>poisthin.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is now defunct. Please try out <code><a href="select_counts.html">select_counts</a></code> and
<code><a href="thin_2group.html">thin_2group</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">poisthin</span><span class="op">(</span>
  <span class="va">mat</span>,
  nsamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>,
  ngene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>,
  gselect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"max"</span>, <span class="st">"random"</span>, <span class="st">"rand_max"</span>, <span class="st">"custom"</span>, <span class="st">"mean_max"</span><span class="op">)</span>,
  gvec <span class="op">=</span> <span class="cn">NULL</span>,
  skip_gene <span class="op">=</span> <span class="fl">0L</span>,
  signal_fun <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span>,
  signal_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fl">0</span>, sd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
  prop_null <span class="op">=</span> <span class="fl">1</span>,
  alpha <span class="op">=</span> <span class="fl">0</span>,
  group_assign <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"frac"</span>, <span class="st">"random"</span>, <span class="st">"cor"</span><span class="op">)</span>,
  group_prop <span class="op">=</span> <span class="fl">0.5</span>,
  corvec <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>mat</dt>
<dd><p>A matrix of count data. The rows index the individuals and
the columns index the genes.</p></dd>
<dt>nsamp</dt>
<dd><p>The number of samples to select from <code>mat</code>.</p></dd>
<dt>ngene</dt>
<dd><p>The number of genes to select from <code>mat</code>.</p></dd>
<dt>gselect</dt>
<dd><p>How should we select the subset of genes? Should we choose
the <code>ngene</code> most median expressed genes (<code>"max"</code>), a random sample
of the genes (<code>"random"</code>), a random sample of the most expressed
genes (<code>"rand_max"</code>), a user-provided list (<code>"custom"</code>), or by maximum
mean expression level (<code>"mean_max"</code>)?
If <code>"custom"</code>, then <code>gvec</code> should be specified. Expression levels
of a gene are measured by median expression across individuals with ties broken
by mean expression.</p></dd>
<dt>gvec</dt>
<dd><p>A logical of length <code>ncol(mat)</code>. A <code>TRUE</code> in position \(i\)
indicates inclusion into the smaller dataset. Hence, <code>sum(gvec)</code> should
equal <code>ngene</code>.</p></dd>
<dt>skip_gene</dt>
<dd><p>The number of maximally expressed genes to skip.
Not used if <code>gselect = "custom"</code>.</p></dd>
<dt>signal_fun</dt>
<dd><p>A function that returns the signal. This should take as
input <code>n</code> for the number of samples to return and then return only
a vector of samples.</p></dd>
<dt>signal_params</dt>
<dd><p>A list of additional arguments to pass to <code>signal_fun</code>.</p></dd>
<dt>prop_null</dt>
<dd><p>The proportion of genes that are null.</p></dd>
<dt>alpha</dt>
<dd><p>If \(b\) is an effect and \(s\) is an empirical standard deviation, then
we model \(b/s^\alpha\) as being exchangeable.</p></dd>
<dt>group_assign</dt>
<dd><p>How should we assign groups? Exactly specifying the
proportion of individuals in each group (<code>"frac"</code>), with a
Bernoulli distribution (<code>"random"</code>), or correlated with latent factors
(<code>"cor"</code>)? If <code>group_assign = "cor"</code>, then you have to specify
<code>corvec</code>. If <code>group_assign = "frac"</code> or
<code>group_assign = "random"</code>, then the proportion of samples in each
group is specified with the <code>group_prop</code> argument.</p></dd>
<dt>group_prop</dt>
<dd><p>The proportion of individuals that are in group 1.
This proportion is deterministic if <code>group_assign = "frac"</code>, and
is the expected proportion if <code>group_assign = "random"</code>. This
argument is not used if <code>group_assign = "cor"</code>.</p></dd>
<dt>corvec</dt>
<dd><p>A vector of correlations. <code>corvec[i]</code> is the correlation
of the latent group assignment vector with the ith latent confounder.
Only used if <code>group_assign = "cor"</code>. This vector is constrained
so that <code>crossprod(corvec) &lt; 1</code>. The number of latent factors
is taken to be the length of corvec. Note that the correlations of the
latent factors with the observed group-assignment vector (instead of the
latent group-assignment vector) will be <code>corvec * sqrt(2 / pi)</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with the following elements:</p><ul><li><p><code>Y</code>: A matrix of altered counts with <code>nsamp</code> rows
       and <code>ngene</code> columns.</p></li>
<li><p><code>X</code>: A design matrix. The first column contains a vector ones (for an
       intercept term) and the second column contains an indicator for group membership.</p></li>
<li><p><code>beta</code>: The approximately true effect sizes of \(log(Y) ~ X\beta\).</p></li>
<li><p><code>corassign</code>: The output from the call to <code><a href="corassign.html">corassign</a></code>.
       Only returned if <code>group_assign = "cor"</code>.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Given a matrix of RNA-seq counts, this function will randomly select two groups of
samples and add signal to a known proportion of the genes. This signal
is the log (base 2) effect size of the group indicator in a linear model.
The user may specify the distribution of the effects.</p>
<p>The Poisson thinning approach first randomly assigns samples to be in one of two groups. Then,
given this assignment, will Binomially sample counts with a sample size of the gene expression
counts and a probability that is a function of the effect size. For details, see
Gerard and Stephens (2021).</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D., and Stephens, M. (2021). "Unifying and Generalizing Methods for Removing Unwanted Variation Based on Negative Controls." <em>Statistica Sinica</em>, 31(3), 1145-1166 doi: <a href="https://doi.org/10.5705/ss.202018.0345" class="external-link">10.5705/ss.202018.0345</a>
.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Simulate data from given matrix of counts</span></span>
<span class="r-in"><span class="co">## In practice, you would obtain Y from a real dataset, not simulate it.</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="va">nsamp</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span class="r-in"><span class="va">ngene</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span class="r-in"><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">nsamp</span> <span class="op">*</span> <span class="va">ngene</span>, lambda <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="va">ngene</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Apply thinning</span></span>
<span class="r-in"><span class="va">poisout</span> <span class="op">&lt;-</span> <span class="fu">poisthin</span><span class="op">(</span>mat           <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>,</span>
<span class="r-in">                    nsamp         <span class="op">=</span> <span class="fl">9</span>,</span>
<span class="r-in">                    ngene         <span class="op">=</span> <span class="fl">999</span>,</span>
<span class="r-in">                    signal_fun    <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span>,</span>
<span class="r-in">                    signal_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fl">0</span>, sd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span class="r-in">                    prop_null     <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> poisthin() is now defunct.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Please use select_counts() and thin_2group() instead.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This message is displayed once per R session</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Dimension of count matrix is smaller.</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">poisout</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]   9 999</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Can verify signal was added by estimating it with lm().</span></span>
<span class="r-in"><span class="va">betahat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">poisout</span><span class="op">$</span><span class="va">Y</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">~</span> <span class="va">poisout</span><span class="op">$</span><span class="va">X</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">poisout</span><span class="op">$</span><span class="va">beta</span>, <span class="va">betahat</span>, xlab <span class="op">=</span> <span class="st">"Coefficients"</span>, ylab <span class="op">=</span> <span class="st">"Estimates"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="fl">2</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span class="r-plt img"><img src="poisthin-1.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

