<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Subsample the rows and columns of a count matrix. — select_counts • seqgendiff</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Subsample the rows and columns of a count matrix. — select_counts"><meta property="og:description" content="It is a good idea to subsample (each iteration) the genes and samples from
a real RNA-seq dataset prior to applying thin_diff
(and related functions) so that your conclusions are not dependent on the
specific structure of your dataset. This function is designed to efficiently
do this for you."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/seqgendiff/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subsample the rows and columns of a count matrix.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/seqgendiff/blob/HEAD/R/select_counts.R" class="external-link"><code>R/select_counts.R</code></a></small>
    <div class="hidden name"><code>select_counts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>It is a good idea to subsample (each iteration) the genes and samples from
a real RNA-seq dataset prior to applying <code><a href="thin_diff.html">thin_diff</a></code>
(and related functions) so that your conclusions are not dependent on the
specific structure of your dataset. This function is designed to efficiently
do this for you.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">select_counts</span><span class="op">(</span></span>
<span>  <span class="va">mat</span>,</span>
<span>  nsamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>,</span>
<span>  ngene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>,</span>
<span>  gselect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"random"</span>, <span class="st">"max"</span>, <span class="st">"mean_max"</span>, <span class="st">"custom"</span><span class="op">)</span>,</span>
<span>  gvec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nskip <span class="op">=</span> <span class="fl">0L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>mat</dt>
<dd><p>A numeric matrix of RNA-seq counts. The rows index the genes
and the columns index the samples.</p></dd>


<dt>nsamp</dt>
<dd><p>The number of samples (columns) to select from <code>mat</code>.</p></dd>


<dt>ngene</dt>
<dd><p>The number of genes (rows) to select from <code>mat</code>.</p></dd>


<dt>gselect</dt>
<dd><p>How should we select the subset of genes? Options include:</p><dl><dt><code>random</code></dt>
<dd><p>Randomly select the genes, with each gene having
      an equal probability of being included in the subsampled matrix.</p></dd>

  <dt><code>max</code></dt>
<dd><p>Choose the <code>ngene</code> most median-expressed genes.
      Ties are broken by mean-expression.</p></dd>

  <dt><code>mean_max</code></dt>
<dd><p>Choose the <code>ngene</code> most mean-expressed
      genes.</p></dd>

  <dt><code>custom</code></dt>
<dd><p>A user-specified list of genes. If
      <code>gselect = "custom"</code> then <code>gvec</code> needs to be
      non-<code>NULL</code>.</p></dd>


</dl></dd>


<dt>gvec</dt>
<dd><p>A logical vector of length <code>nrow(mat)</code>. A <code>TRUE</code>
in position \(i\) indicates inclusion into the smaller dataset.
Hence, <code>sum(gvec)</code> should equal <code>ngene</code>.</p></dd>


<dt>filter_first</dt>
<dd><p>Should we first filter genes by the method of
Chen et al. (2016) (<code>TRUE</code>) or not (<code>FALSE</code>)? If
<code>TRUE</code> then the <code>edgeR</code> package should be installed.</p></dd>


<dt>nskip</dt>
<dd><p>The number of median-maximally expressed genes to skip.
Not used if <code>gselect = "custom"</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A numeric matrix, which is a <code>ngene</code> by <code>nsamp</code> sub-matrix
    of <code>mat</code>. If <code>rownames(mat)</code> is <code>NULL</code>, then the
    row names of the returned matrix are the indices in <code>mat</code> of the
    selected genes. If <code>colnames(mat)</code> is <code>NULL</code>, then the
    column names of the returned matrix are the indices in <code>mat</code> of
    the selected samples.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The samples (columns) are chosen randomly, with each sample having
an equal probability of being in the sub-matrix. The genes are selected
according to one of four schemes (see the description of the <code>gselect</code>
argument).</p>
<p>If you have edgeR installed, then some functionality is provided for
filtering out the lowest expressed genes prior to applying subsampling
(see the <code>filter_first</code> argument).
This filtering scheme is described in Chen et al. (2016).
If you want more control over this filtering, you should use
the <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></code> function from edgeR directly. You
can install edgeR by following instructions at
<a href="https://doi.org/10.18129/B9.bioc.edgeR" class="external-link">doi:10.18129/B9.bioc.edgeR</a>
.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Chen, Yunshun, Aaron TL Lun, and Gordon K. Smyth. "From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline." <em>F1000Research</em> 5 (2016). <a href="https://doi.org/10.12688/f1000research.8987.2" class="external-link">doi:10.12688/f1000research.8987.2</a>
.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Simulate data from given matrix of counts</span></span></span>
<span class="r-in"><span><span class="co">## In practice, you would obtain mat from a real dataset, not simulate it.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">n</span>   <span class="op">&lt;-</span> <span class="fl">100</span></span></span>
<span class="r-in"><span><span class="va">p</span>   <span class="op">&lt;-</span> <span class="fl">1000</span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">n</span> <span class="op">*</span> <span class="va">p</span>, lambda <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="va">p</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Subsample the matrix, then feed it into a thinning function</span></span></span>
<span class="r-in"><span><span class="va">submat</span> <span class="op">&lt;-</span> <span class="fu">select_counts</span><span class="op">(</span>mat <span class="op">=</span> <span class="va">mat</span>, nsamp <span class="op">=</span> <span class="fl">10</span>, ngene <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">thout</span>  <span class="op">&lt;-</span> <span class="fu"><a href="thin_2group.html">thin_2group</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">submat</span>, prop_null <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The rownames and colnames (if NULL in mat) tell you which genes/samples</span></span></span>
<span class="r-in"><span><span class="co">## were selected.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">submat</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1] "12"  "22"  "40"  "57"  "58"  "70"  "79"  "110" "113" "124" "125" "137"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [13] "141" "142" "153" "186" "200" "236" "237" "239" "251" "254" "257" "259"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [25] "260" "285" "290" "293" "321" "329" "333" "342" "343" "356" "359" "369"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [37] "383" "394" "401" "420" "421" "435" "437" "467" "468" "476" "481" "490"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [49] "492" "496" "507" "519" "520" "549" "588" "597" "616" "620" "641" "647"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [61] "648" "652" "655" "657" "659" "683" "704" "714" "717" "728" "729" "732"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [73] "734" "740" "745" "764" "766" "773" "778" "779" "782" "789" "801" "816"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [85] "836" "845" "848" "849" "857" "863" "877" "904" "913" "919" "930" "947"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [97] "952" "971" "992" "999"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">submat</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "3"  "9"  "13" "14" "40" "80" "84" "86" "94" "99"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

