<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>seqgendiff: RNA-Seq Generation/Modification for Simulation — seqgendiff-package • seqgendiff</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="seqgendiff: RNA-Seq Generation/Modification for Simulation — seqgendiff-package"><meta property="og:description" content="This package is designed to take real RNA-seq data and alter it by
adding a known amount of signal. You can then use this modified dataset
in simulation studies for differential expression analysis, factor
analysis, confounder adjustment, or library size adjustment. The
advantage of this way of simulating data is that you can see how
your method behaves when the simulated data exhibit common
(and annoying) features of real data. For example, in the real world
data are not normally (or negative binomially) distributed and
unobserved confounding is a major issue. This package will simulate
data that exhibit these characteristics. The methods used in this
package are described in detail in Gerard (2020)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/seqgendiff/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>seqgendiff: RNA-Seq Generation/Modification for Simulation</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/seqgendiff/blob/HEAD/R/seqgendiff.R" class="external-link"><code>R/seqgendiff.R</code></a></small>
    <div class="hidden name"><code>seqgendiff-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This package is designed to take real RNA-seq data and alter it by
adding a known amount of signal. You can then use this modified dataset
in simulation studies for differential expression analysis, factor
analysis, confounder adjustment, or library size adjustment. The
advantage of this way of simulating data is that you can see how
your method behaves when the simulated data exhibit common
(and annoying) features of real data. For example, in the real world
data are not normally (or negative binomially) distributed and
unobserved confounding is a major issue. This package will simulate
data that exhibit these characteristics. The methods used in this
package are described in detail in Gerard (2020).</p>
    </div>


    <div id="seqgendiff-functions">
    <h2>seqgendiff Functions</h2>
    

<dl><dt><code><a href="select_counts.html">select_counts</a>()</code></dt>
<dd><p>Subsample the columns and rows
      of a real RNA-seq count matrix. You would then feed this sub-matrix
      into one of the thinning functions below.</p></dd>

  <dt><code><a href="thin_diff.html">thin_diff</a>()</code></dt>
<dd><p>The function most users should
      be using for general-purpose binomial thinning. For the special
      applications of the two-group model or library/gene thinning, see
      the functions listed below.</p></dd>

  <dt><code><a href="thin_2group.html">thin_2group</a>()</code></dt>
<dd><p>The specific application of
      thinning in the two-group model.</p></dd>

  <dt><code><a href="thin_lib.html">thin_lib</a>()</code></dt>
<dd><p>The specific application of
      library size thinning.</p></dd>

  <dt><code><a href="thin_gene.html">thin_gene</a>()</code></dt>
<dd><p>The specific application of
      total gene expression thinning.</p></dd>

  <dt><code><a href="thin_all.html">thin_all</a>()</code></dt>
<dd><p>The specific application of thinning
      all counts.</p></dd>

  <dt><code><a href="effective_cor.html">effective_cor</a>()</code></dt>
<dd><p>Returns an estimate of the actual
      correlation between surrogate variables and a user-specified
      design matrix.</p></dd>

  <dt><code><a href="ThinDataToSummarizedExperiment.html">ThinDataToSummarizedExperiment</a>()</code></dt>
<dd><p>Converts a
      ThinData object to a SummarizedExperiment object.</p></dd>

  <dt><code><a href="ThinDataToDESeqDataSet.html">ThinDataToDESeqDataSet</a>()</code></dt>
<dd><p>Converts a
      ThinData object to a DESeqDataSet object.</p></dd>


</dl></div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D (2020). "Data-based RNA-seq simulations by binomial thinning." <em>BMC Bioinformatics</em>. 21(1), 206. <a href="https://doi.org/10.1186/s12859-020-3450-9" class="external-link">doi:10.1186/s12859-020-3450-9</a>
.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://github.com/dcgerard/seqgendiff" class="external-link">https://github.com/dcgerard/seqgendiff</a></p></li>
<li><p>Report bugs at <a href="https://github.com/dcgerard/seqgendiff/issues" class="external-link">https://github.com/dcgerard/seqgendiff/issues</a></p></li>
</ul></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

