<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Base binomial thinning function. — thin_base • seqgendiff</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Base binomial thinning function. — thin_base"><meta property="og:description" content="Given a matrix of counts (\(Y\)) where \(log_2(E[Y]) = Q\),
a design matrix (\(X\)), and a matrix of coefficients (\(B\)),
thin_diff will generate a new matrix of counts such that
\(log_2(E[Y]) = BX' + u1' + Q\), where \(u\) is some vector
of intercept coefficients. This function is used by all other
thinning functions. The method is
described in detail in Gerard (2020)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="http://github.com/dcgerard/seqgendiff/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Base binomial thinning function.</h1>
    <small class="dont-index">Source: <a href="http://github.com/dcgerard/seqgendiff/blob/HEAD/R/genthin.R" class="external-link"><code>R/genthin.R</code></a></small>
    <div class="hidden name"><code>thin_base.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a matrix of counts (\(Y\)) where \(log_2(E[Y]) = Q\),
a design matrix (\(X\)), and a matrix of coefficients (\(B\)),
<code>thin_diff</code> will generate a new matrix of counts such that
\(log_2(E[Y]) = BX' + u1' + Q\), where \(u\) is some vector
of intercept coefficients. This function is used by all other
thinning functions. The method is
described in detail in Gerard (2020).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">thin_base</span><span class="op">(</span><span class="va">mat</span>, <span class="va">designmat</span>, <span class="va">coefmat</span>, relative <span class="op">=</span> <span class="cn">TRUE</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"thin"</span>, <span class="st">"mult"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>mat</dt>
<dd><p>A numeric matrix of RNA-seq counts. The rows index the genes and
the columns index the samples.</p></dd>
<dt>designmat</dt>
<dd><p>A design matrix. The rows index the samples and the columns
index the variables. The intercept should <em>not</em> be included.</p></dd>
<dt>coefmat</dt>
<dd><p>A matrix of coefficients. The rows index the genes and the
columns index the samples.</p></dd>
<dt>relative</dt>
<dd><p>A logical. Should we apply relative thinning (<code>TRUE</code>)
or absolute thinning (<code>FALSE</code>). Only experts should change
the default.</p></dd>
<dt>type</dt>
<dd><p>Should we apply binomial thinning (<code>type = "thin"</code>) or
just naive multiplication of the counts (<code>type = "mult"</code>).
You should always have this set to <code>"thin"</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A matrix of new RNA-seq read-counts. This matrix has the signal
    added from <code>designmat</code> and <code>coefmat</code>.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D (2020). "Data-based RNA-seq simulations by binomial thinning." <em>BMC Bioinformatics</em>. 21(1), 206. doi: <a href="https://doi.org/10.1186/s12859-020-3450-9" class="external-link">10.1186/s12859-020-3450-9</a>.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index">
<dl><dt><code><a href="select_counts.html">select_counts</a></code></dt>
<dd><p>For subsampling the rows and columns
      of your real RNA-seq count matrix prior to applying binomial thinning.</p></dd>

  <dt><code><a href="thin_diff.html">thin_diff</a></code></dt>
<dd><p>For the function most users should
      be using for general-purpose binomial thinning.</p></dd>

  <dt><code><a href="thin_2group.html">thin_2group</a></code></dt>
<dd><p>For the specific application of
      thinning in the two-group model.</p></dd>

  <dt><code><a href="thin_lib.html">thin_lib</a></code></dt>
<dd><p>For the specific application of
      library size thinning.</p></dd>

  <dt><code><a href="thin_gene.html">thin_gene</a></code></dt>
<dd><p>For the specific application of
      total gene expression thinning.</p></dd>

  <dt><code><a href="thin_all.html">thin_all</a></code></dt>
<dd><p>For the specific application of
      thinning all counts uniformly.</p></dd>


</dl></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Simulate data from given matrix of counts</span></span>
<span class="r-in"><span class="co">## In practice, you would obtain Y from a real dataset, not simulate it.</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="va">nsamp</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span class="r-in"><span class="va">ngene</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span class="r-in"><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">nsamp</span> <span class="op">*</span> <span class="va">ngene</span>, lambda <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="va">ngene</span><span class="op">)</span></span>
<span class="r-in"><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="va">nsamp</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">0</span>, length.out <span class="op">=</span> <span class="va">ngene</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">Ynew</span> <span class="op">&lt;-</span> <span class="fu">thin_base</span><span class="op">(</span>mat <span class="op">=</span> <span class="va">Y</span>, designmat <span class="op">=</span> <span class="va">X</span>, coefmat <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Demonstrate how the log2 effect size is B</span></span>
<span class="r-in"><span class="va">Bhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coefficients</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">Ynew</span><span class="op">)</span><span class="op">)</span> <span class="op">~</span> <span class="va">X</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="st">"X"</span>, <span class="op">]</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">B</span>, <span class="va">Bhat</span>, xlab <span class="op">=</span> <span class="st">"Coefficients"</span>, ylab <span class="op">=</span> <span class="st">"Coefficient Estimates"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span class="r-plt img"><img src="thin_base-1.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

