<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Binomial thinning in the two-group model. — thin_2group • seqgendiff</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Binomial thinning in the two-group model. — thin_2group"><meta property="og:description" content="Given a matrix of real RNA-seq counts, this function will
randomly assign samples to one of two groups, draw
the log2-fold change in expression between two groups for each gene,
and add this signal to the RNA-seq counts matrix. The user may specify
the proportion of samples in each group, the proportion of null genes
(where the log2-fold change is 0),
and the signal function. This is a specific application of the
general binomial thinning approach implemented in thin_diff."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/seqgendiff/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Binomial thinning in the two-group model.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/seqgendiff/blob/HEAD/R/genthin.R" class="external-link"><code>R/genthin.R</code></a></small>
    <div class="hidden name"><code>thin_2group.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a matrix of real RNA-seq counts, this function will
randomly assign samples to one of two groups, draw
the log2-fold change in expression between two groups for each gene,
and add this signal to the RNA-seq counts matrix. The user may specify
the proportion of samples in each group, the proportion of null genes
(where the log2-fold change is 0),
and the signal function. This is a specific application of the
general binomial thinning approach implemented in <code><a href="thin_diff.html">thin_diff</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">thin_2group</span><span class="op">(</span></span>
<span>  <span class="va">mat</span>,</span>
<span>  prop_null <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  signal_fun <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span>,</span>
<span>  signal_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fl">0</span>, sd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  group_prop <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  corvec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  permute_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hungarian"</span>, <span class="st">"marriage"</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"thin"</span>, <span class="st">"mult"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>mat</dt>
<dd><p>A numeric matrix of RNA-seq counts. The rows index the genes and
the columns index the samples.</p></dd>


<dt>prop_null</dt>
<dd><p>The proportion of genes that are null.</p></dd>


<dt>signal_fun</dt>
<dd><p>A function that returns the signal. This should take as
input <code>n</code> for the number of samples to return and then return only
a vector of samples. Additional parameters may be passed through
<code>signal_params</code>.</p></dd>


<dt>signal_params</dt>
<dd><p>A list of additional arguments to pass to
<code>signal_fun</code>.</p></dd>


<dt>group_prop</dt>
<dd><p>The proportion of individuals that are in group 1.</p></dd>


<dt>corvec</dt>
<dd><p>A vector of target correlations. <code>corvec[i]</code> is the
target correlation of the latent group assignment vector with the
<code>i</code>th surrogate variable. The default is to set this to <code>NULL</code>,
in which case group assignment is made independently of any
unobserved confounding.</p></dd>


<dt>alpha</dt>
<dd><p>The scaling factor for the signal distribution from
Stephens (2016). If \(x_1, x_2, ..., x_n\) are drawn from
<code>signal_fun</code>, then the signal is set to
\(x_1 s_1^{\alpha}, x_2 s_2^{\alpha}, ..., x_n s_n^{\alpha}\), where
\(s_g\) is the empirical standard deviation of gene \(g\).
Setting this to <code>0</code> means that the effects are exchangeable, setting
this to <code>1</code> corresponds to the p-value prior of
Wakefield (2009). You would rarely set this to anything but <code>0</code>
or <code>1</code>.</p></dd>


<dt>permute_method</dt>
<dd><p>Should we use the Gale-Shapley algorithm
for stable marriages (<code>"marriage"</code>) (Gale and Shapley, 1962)
as implemented in the matchingR package, or the Hungarian algorithm
(Papadimitriou and Steiglitz, 1982) (<code>"hungarian"</code>)
as implemented in the clue package (Hornik, 2005)? The
Hungarian method almost always works better, so is the default.</p></dd>


<dt>type</dt>
<dd><p>Should we apply binomial thinning (<code>type = "thin"</code>) or
just naive multiplication of the counts (<code>type = "mult"</code>).
You should always have this set to <code>"thin"</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list-like S3 object of class <code>ThinData</code>.
Components include some or all of the following:</p><dl><dt><code>mat</code></dt>
<dd><p>The modified matrix of counts.</p></dd>

  <dt><code>designmat</code></dt>
<dd><p>The design matrix of variables used to simulate
      signal. This is made by column-binding <code>design_fixed</code> and the
      permuted version of <code>design_perm</code>.</p></dd>

  <dt><code>coefmat</code></dt>
<dd><p>A matrix of coefficients corresponding to
      <code>designmat</code>.</p></dd>

  <dt><code>design_obs</code></dt>
<dd><p>Additional variables that should be included in
      your design matrix in downstream fittings. This is made by
      column-binding the vector of 1's with <code>design_obs</code>.</p></dd>

  <dt><code>sv</code></dt>
<dd><p>A matrix of estimated surrogate variables. In simulation
      studies you would probably leave this out and estimate your own
      surrogate variables.</p></dd>

  <dt><code>cormat</code></dt>
<dd><p>A matrix of target correlations between the
      surrogate variables and the permuted variables in the design matrix.
      This might be different from the <code>target_cor</code> you input because
      we pass it through <code><a href="fix_cor.html">fix_cor</a></code> to ensure
      positive semi-definiteness of the resulting covariance matrix.</p></dd>

  <dt><code>matching_var</code></dt>
<dd><p>A matrix of simulated variables used to
      permute <code>design_perm</code> if the <code>target_cor</code> is not
      <code>NULL</code>.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>The specific application of binomial thinning to the two-group model was
used in Gerard and Stephens (2018) and Gerard and Stephens (2021). This is
a specific case of the general method described in Gerard (2020).</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gale, David, and Lloyd S. Shapley. "College admissions and the stability of marriage." <em>The American Mathematical Monthly</em> 69, no. 1 (1962): 9-15. <a href="https://doi.org/10.1080/00029890.1962.11989827" class="external-link">doi:10.1080/00029890.1962.11989827</a>
.</p></li>
<li><p>Gerard, D., and Stephens, M. (2021). "Unifying and Generalizing Methods for Removing Unwanted Variation Based on Negative Controls." <em>Statistica Sinica</em>, 31(3), 1145-1166 <a href="https://doi.org/10.5705/ss.202018.0345" class="external-link">doi:10.5705/ss.202018.0345</a>
.</p></li>
<li><p>David Gerard and Matthew Stephens (2018). "Empirical Bayes shrinkage and false discovery rate estimation, allowing for unwanted variation." <em>Biostatistics</em>, <a href="https://doi.org/10.1093/biostatistics/kxy029" class="external-link">doi:10.1093/biostatistics/kxy029</a>
.</p></li>
<li><p>Gerard, D (2020). "Data-based RNA-seq simulations by binomial thinning." <em>BMC Bioinformatics</em>. 21(1), 206. <a href="https://doi.org/10.1186/s12859-020-3450-9" class="external-link">doi:10.1186/s12859-020-3450-9</a>
.</p></li>
<li><p>Hornik K (2005). "A CLUE for CLUster Ensembles." <em>Journal of Statistical Software</em>, 14(12). <a href="https://doi.org/10.18637/jss.v014.i12" class="external-link">doi:10.18637/jss.v014.i12</a>
. <a href="https://doi.org/10.18637/jss.v014.i12" class="external-link">doi:10.18637/jss.v014.i12</a>
.</p></li>
<li><p>C. Papadimitriou and K. Steiglitz (1982), Combinatorial Optimization: Algorithms and Complexity. Englewood Cliffs: Prentice Hall.</p></li>
<li><p>Stephens, Matthew. "False discovery rates: a new deal." <em>Biostatistics</em> 18, no. 2 (2016): 275-294. <a href="https://doi.org/10.1093/biostatistics/kxw041" class="external-link">doi:10.1093/biostatistics/kxw041</a>
.</p></li>
<li><p>Wakefield, Jon. "Bayes factors for genome-wide association studies: comparison with P-values." <em>Genetic epidemiology</em> 33, no. 1 (2009): 79-86. <a href="https://doi.org/10.1002/gepi.20359" class="external-link">doi:10.1002/gepi.20359</a>
.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index">
<dl><dt><code><a href="select_counts.html">select_counts</a></code></dt>
<dd><p>For subsampling the rows and columns
      of your real RNA-seq count matrix prior to applying binomial thinning.</p></dd>

  <dt><code><a href="thin_diff.html">thin_diff</a></code></dt>
<dd><p>For the more general thinning approach.</p></dd>

  <dt><code><a href="ThinDataToSummarizedExperiment.html">ThinDataToSummarizedExperiment</a></code></dt>
<dd><p>For converting a
      ThinData object to a SummarizedExperiment object.</p></dd>

  <dt><code><a href="ThinDataToDESeqDataSet.html">ThinDataToDESeqDataSet</a></code></dt>
<dd><p>For converting a
      ThinData object to a DESeqDataSet object.</p></dd>


</dl></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Simulate data from given matrix of counts</span></span></span>
<span class="r-in"><span><span class="co">## In practice, you would obtain Y from a real dataset, not simulate it.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">nsamp</span> <span class="op">&lt;-</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="va">ngene</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span></span>
<span class="r-in"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">nsamp</span> <span class="op">*</span> <span class="va">ngene</span>, lambda <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="va">ngene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">thinout</span> <span class="op">&lt;-</span> <span class="fu">thin_2group</span><span class="op">(</span>mat           <span class="op">=</span> <span class="va">Y</span>,</span></span>
<span class="r-in"><span>                       prop_null     <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>                       signal_fun    <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Exponential.html" class="external-link">rexp</a></span>,</span></span>
<span class="r-in"><span>                       signal_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rate <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## 90 percent of genes are null</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">thinout</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">6</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.9</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Check the estimates of the log2-fold change</span></span></span>
<span class="r-in"><span><span class="va">Ynew</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">thinout</span><span class="op">$</span><span class="va">mat</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">X</span>    <span class="op">&lt;-</span> <span class="va">thinout</span><span class="op">$</span><span class="va">designmat</span></span></span>
<span class="r-in"><span><span class="va">Bhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">Ynew</span> <span class="op">~</span> <span class="va">X</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="st">"X"</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">thinout</span><span class="op">$</span><span class="va">coefmat</span>,</span></span>
<span class="r-in"><span>     <span class="va">Bhat</span>,</span></span>
<span class="r-in"><span>     xlab <span class="op">=</span> <span class="st">"log2-fold change"</span>,</span></span>
<span class="r-in"><span>     ylab <span class="op">=</span> <span class="st">"Estimated log2-fold change"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="thin_2group-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

